extensions/lifecycle/LifecycleQueuePopulator.js

```diff
@@ -0,0 +1,136 @@
+const { usersBucket, mpuBucketPrefix } = require('arsenal').constants;
+const QueuePopulatorExtension =
+    require('../../lib/queuePopulator/QueuePopulatorExtension');
+
+class LifecycleQueuePopulator extends QueuePopulatorExtension {
+
+    /**
+     * Pre-create the zookeeper node path for bucket lifecycle nodes, if
+     * necessary.
+     * @param {Function} cb - The callback function.
+     * @return {undefined}
+     */
+    _setup(cb) {
+        return this.zkClient.getData(this.zkPath, err => {
+            if (err) {
+                if (err.name !== 'NO_NODE') {
+                    this.log.error('Could not get lifecycle nodes', {
+                        method: 'LifecycleQueuePopulator._setup',
+                        error: err,
+                    });
+                    return cb(err);
+                }
+                return this.zkClient.mkdirp(this.zkPath, err => {
+                    if (err) {
+                        this.log.error('Could not create path in zookeeper', {
+                            method: 'LifecycleQueuePopulator._setup',
+                            zkPath: this.zkPath,
+                            error: err,
+                        });
+                        return cb(err);
+                    }
+                    return cb();
+                });
+            }
+            return cb();
+        });
+    }
+
+    /**
+     * Create a new zookeeper node for any bucket that has a lifecycle
+     * configuration. Remove the node if the bucket is deleted or the
+     * configuration is deleted.
+     * @param {Object} entry - The record log entry from metadata.
+     * @return {undefined}
+     */
+    _updateZookeeper(entry) {
+        const value = JSON.parse(entry.value);
+        const path = `${this.zkPath}/${value.owner}:${value.name}`;
+        return this.zkClient.exists(path, (err, stat) => {
```

Comment:
It should not be necessary to call `exists` before a `remove` unless we need to look into the `stat` result: if `remove` returns `ZNONODE` error, we can ignore it because it means the node doesn't exist, saving one zookeeper operation.

Reply:
Updated.